### PR TITLE
Refactored hex,added validation checks. Issue #175

### PR DESCRIPTION
**Pull request by Kurios**
Currently working on map validation fix for Issue #175, to prevent maps with terrain like woods:0 from becoming an issue.

### CODE REVIEW

**Comment by Kurios** [comment]
Pull request can be merged now. 

WIll validate maps against a small set of rules, and patch the woods:0 problem now. If a random map would violate rules, it picks another map.

**Comment by Akjosch** [comment]
Small inconsistency: Why is woods "fixing" its terrain level, but nothing else does?

While talking implementing limits and auto-fixes, roads should be limited to values of 1 and 2 (with 1 being the default when outside of those values), with a value of 2 a legacy issue and discouraged.

**Comment by Kurios** [comment]
want me to force all roads to 1 then?

And I only changed woods because that's the current known problem.

**Comment by Akjosch** [comment]
Roads should be 1 or 2 for now. Limiting them to just 1 would break too many things at the moment.

What's the current known problem with the woods exactly? Because if it is something in the provided maps or the terrain generator, the solution is to fix *them*.

**Comment by Kurios** [comment]
some of the the current maps have woods:0, which is allowing infantry free movement through them. ( #175 ). Starting a primitive map sanity check was one of the solutions that came out of yesterday's discussion, in order to prevent these breaking maps from hitting. Then this afternoon, it was requested that we don't reject the wood:0, but patch it in place, which was why that was done differently.

**Comment by Akjosch** [comment]
Provided maps are fixed.

**Comment by beerockxs** [comment]
Please format your code to use 4 spaces for indent and to have a space before opening braces.

**Comment by Kurios** [comment]
@beerockxs: Whomever is in charge of formatting should come up with a lint or some sort of formatting tool (like the eclipse code formatter) so the formatting reqs are in code (And if possible, automated)

File: megamek/src/megamek/common/Board.java [hunk 1]
```diff
@@ -897,15 +897,32 @@ public void load(InputStream is) {
...
+	}
+    
+    private boolean isValid(IHex[] data) {
+		//Search for black-listed hexes
+    	for(IHex hex: data) {
+    		if(!hex.isValid()) return false;	
+    	}
+		return true;
+	}
+
```
Comment (arlith) [inline megamek/src/megamek/common/Board.java]: It seems like there's unnecessary code duplication here.  Why does `isValid(0)` not just call `isValid(1)`?  Why does `isValid(1)` not check for a null hex but `isValid(0)` does?

**Comment by arlith** [comment]
Another thought that occurred to me... There are some terrain types that require a combination of terrains.  For example, buildings.  There are at least three `Terrain`'s that are required: building, bldg_cf, bldg_elev.  There may be another I'm forgetting.  But, we should validate that in these cases, all of the required `Terrains` are defined.

**Comment by Kurios** [comment]
Arlith: Mind if I put this on back-burner, and start a major formatting only change to make this all cleaner?

**Comment by arlith** [comment]
I think this is pretty close to merging in, and just needs a few small changes and I'll merge it in.  In addition to the comment I made, it would also be nice to be able to get a report of all the boards currently in MM.  Some type of app like in megamek.test (say, MechSetTest).  I could probably write this app myself though.  If you make the one change I commented on, I'll probably merge this in.

**Pull request merged**